### ISSUE TITLE
Get the current window scene from SwiftUI

### ISSUE THREAD

**Issue by aheze** [aheze/Popovers#3]
If you want support multiple screens, **and** use frame tags or `Popovers.popover(tagged:)`, you need to specify the window scene. The problem is, getting a view's window scene in SwiftUI is weird.

My current workaround (involves 0.5 second hardcoded delay): https://github.com/aheze/Popovers/blob/5fcaa9d9eb2ed077cd43e323b5a772a04bd6e1be/Sources/PopoverFrameTag.swift#L67-L114

You can then use this in your app's root view via `injectWindowScene`.

https://github.com/aheze/Popovers/blob/8e22ac703bb847463cd863c22bfd5a3047be5c00/Sources/PopoverFrameTag.swift#L130-L138

```swift
WindowGroup {
    ContentView()
        .injectWindowScene() /// Make the window scene available to all subviews. Not ideal, but it works (usually).
}

struct ContentView: View {
    @EnvironmentObject var windowSceneModel: WindowSceneModel

    /// ... 

    Text("Hello").frameTag("Your Frame Tag Name", in: windowSceneModel.windowScene)
}
```

The hardcoded delay especially is annoying. Typing out `@EnvironmentObject var windowSceneModel: WindowSceneModel` every time you want to access the window scene is ok, but it's still kind of long.

**Does anyone know how to get a view's window in SwiftUI?** Preferably, something like this:

```swift
struct ContentView: View {
    var body: some View {
        Text("Hello")
            .frameTag("Your Frame Tag Name")
    }
}
public extension View {
    func frameTag(_ tag: String) -> some View {
        let windowScene = theWindowSceneThatHostsThisView /// Something like this? So no more need to specify the window scene manually.
        return self.modifier(FrameTagModifier(tag: tag, windowScene: windowScene))
    }
}
```

**Comment by sans-connaissance** [comment]
Would something like this work?


```
struct WindowSceneReader: UIViewRepresentable {
    
    /// A closure that's called when the window is found.
    var found: ((UIWindowScene?) -> Void)
    
    func makeUIView(context: Context) -> UIView {
        let view = UIView()

        /**
         The 0.5 second delay is needed to wait until the window is first initialized.
         However, it's hardcoded. Does anyone know how to work around this?
         */
/// Use windowSceneDelegate to pass an initialized windowScene to found(...)
/// From https://stackoverflow.com/questions/62652555/how-to-get-uiwindow-value-in-ios-14-main-file
        var window: UIWindow? {
            guard let scene = UIApplication.shared.connectedScenes.first,
                  let windowSceneDelegate = scene.delegate as? UIWindowSceneDelegate,
                  let window = windowSceneDelegate.window else {
                return nil
            }
            return window
        }
        
        DispatchQueue.main.async {
            found(window?.windowScene)
        }
        
        return view
    }
    
    func updateUIView(_ uiView: UIView, context: Context) { }
}

```

**Comment by aheze** [comment]
@sans-connaissance thanks, that would indeed get the main window. But I need the *current* window that hosts the view. Around 99% of the time this is the same as the main window... but if your app supports multiple windows (iPadOS), it will be different. Should have been more clear, sorry.

```swift
/// Multiple copies of this view can be created if your app supports multiple windows.
struct ContentView: View {
    var body: some View {
        Text("Hello!") /// <- what window scene is this embedded in?
    }
}
```

But anyway, I think ShezHsky's #5 PR fixes this issue. I'm still checking over it to see if it will work out, but will update this thread when I'm done.

https://github.com/aheze/Popovers/pull/5/files#diff-a1c21266189878135a0f3c0581755fcd2da315bfa06f0e3c1d8de478d85d48f3

**Comment by sans-connaissance** [comment]
Thanks for taking a look! This issue and your comment helped me to better understand the window and view relationship in SwiftUI projects.

**Comment by aheze** [comment]
Fixed in #5.

**Comment by rijieli** [comment]
WindowSceneReader is very useful on visionOS when you have multiple windows and want to lock the window resize ratio.
